### PR TITLE
destination: Remove support for IP Queries in `Get` API

### LINKED ISSUE #5246: Drop support for IP queries in destination.Get
The Destination controller's `Get` API accepts queries for both fully qualified authorities as well as for IP addresses.  However, as of stable-2.9.0, the Linkerd proxy will never send IP address queries to the `Get` API.  In order to maintain one version of backwards compatibility between the proxy and the control plane, the destination control should continue to support IP address Get queries in stable-2.10.  However, we should drop support for IP address queries to Destination.Get in stable-2.11 which will allow us to simplify the IP watching logic.

In contrast, the `GetProfiles` API must continue to support fully qualified authority queries (used by the inbound proxy) and IP address queries (used by the outbound proxy).

### PR DESCRIPTION
**Pull request by Pothulapati**
Fixes #5246

This PR updates the destination to report an error when `Get`
is called for IP Queries. As the issue mentions, The proxies
are not using this API anymore and it helps to simplify and
remove unecessary logic.

This change is aimed for `stable-2.11`, and depends on
https://github.com/linkerd/linkerd2-proxy/pull/972

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

### CODE REVIEW

**Comment by Pothulapati** [comment]
I was not able to remove more logic from `ipWatcher` as the indexers, etc underlying state are used by other methods i.e `GetPod`, `GetSvcID`, etc which are then used in the `GetProfile` method.

Feel free to suggest if there are any ways to further simplify this logic.

cc: @adleong

**Review by kleimkuhler — COMMENTED** [review]
So the IP watcher is used for `GetSvcID` and `GetPod` now. However, both of those functions only use the `k8sAPI` field which we have access to in the server. We should change those from methods on `IPWatcher` to methods on `Server`.

`NewIPWatcher` has a side effect which adds the `podIPIndex` and `hostIPIndex` indexers to the `k8sAPI` that is passed in. Adding those indexers should probably move to `Server`'s initialization. After doing that, I think we should be able to fully get rid of `IPWatcher`.

File: controller/api/destination/server.go [hunk 1]
```diff
@@ -352,6 +351,127 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
...
+
+	return !podTerminating && !podTerminated
+}
+
+// podToAddressSet converts a Pod spec into a set of Addresses.
+func podToAddressSet(k8sAPI *k8s.API, pod *corev1.Pod) *watcher.AddressSet {
+	ownerKind, ownerName := k8sAPI.GetOwnerKindAndName(context.Background(), pod, true)
+	return &watcher.AddressSet{
+		Addresses: map[watcher.PodID]watcher.Address{
+			watcher.PodID{
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:460]: Linter is suggesting this as I review 😬:
```suggestion
			{
```

File: controller/api/destination/server.go [hunk 2]
```diff
@@ -352,6 +351,127 @@ func (s *server) GetProfile(dest *pb.GetDestination, stream pb.Destination_GetPr
...
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			}: watcher.Address{
```
Comment (kleimkuhler) [inline controller/api/destination/server.go:463]: Same here:
```suggestion
			}: {
```

**Review by kleimkuhler — COMMENTED** [review]
Overall this looks good. I think moving the watcher setup into a function makes sense so that it's more obvious that we are adding indexers as part of its initialization.

I was going to suggest making `getSvcID` and `getPod` as methods on `Server`, but since we don't use any other fields from the server I think they are okay as regular functions. That being said, it may be worth making them methods on `k8s.API` in the `controller` package. That way we can keep functions like this in a singular place.

Old:
```
svcID, err := getSvcID(s.k8sAPI, ip.String(), log)
```

New:
```
svcID, err := s.k8sAPI.getSvcID(ip.String(), log)
```

**Review by adleong — APPROVED** [review]
👑 💇

**Comment by Pothulapati** [comment]
> That being said, it may be worth making them methods on k8s.API in the controller package. 

I'm not sure if it is the right way to do it for the following reasons:

- It would introduce a lot of the indexing logic and types into `controller/api.go`
- This also introduces circular dependencies and makes the logic complicated.

How about we keep this logic in the same pkg but into a different file, so that `server.go` is cleaner?

**Review by kleimkuhler — APPROVED** [review]
I'm okay with it is as is then. I don't think introducing a new file will necessarily make it easier to read or understand. It'd be nice if we had a way to track these functions that are based off having a `k8sAPI`—and making them methods does seem ideal—but if it brings in a lot of extra dependencies and all then I also don't think that's the right solution.

**Pull request merged**